crates/rwasm/machine/src/alu/rotate/mod.rs

```diff
@@ -0,0 +1,522 @@
+#[derive(AlignedBorrow, Default, Debug, Clone, Copy)]
+#[repr(C)]
+pub struct RotateCols<T> {
+    /// The program counter.
+    pub pc: T,
+    /// The final result of the rotation.
+    pub a: Word<T>,
+    /// The 32-bit value to be rotated.
+    pub b: Word<T>,
+    /// The 32-bit rotation amount.
+    pub c: Word<T>,
+
+    /// The rotation amount, masked to 5 bits (c & 0x1F).
+    pub c_masked: Word<T>,
+    /// The inverse rotation amount for the left shift (32 - c_masked).
+    pub c_inverse: Word<T>,
+
+    /// The result of the right shift (contribution from the right side).
+    pub right_shifted: Word<T>,
+    /// The result of the left shift (contribution from the left side).
+    pub left_shifted: Word<T>,
+
+    /// Instruction flag: set to 1 if the instruction is I32Rotr.
+    pub is_rotr: T,
+    /// Instruction flag: set to 1 if the instruction is I32Rotl.
+    pub is_rotl: T,
+}
+
+/// A chip that implements alu operations for the opcodes I32Rotl and I32Rotr.
+#[derive(Default)]
+pub struct RotateChip;
+
+impl RotateChip {
+    /// Create a row from an event.
+    fn event_to_row<F: PrimeField>(
+        &self,
+        event: &AluEvent,
+        cols: &mut RotateCols<F>,
+        blu: &mut impl ByteRecord,
+    ) {
+        // --- 1) Basic columns: copy CPU operands/results into the trace row.
+        cols.pc = F::from_canonical_u32(event.pc);
+
+        cols.a = Word::from(event.a);
+        cols.b = Word::from(event.b);
+        cols.c = Word::from(event.c);
+
+        let k = (event.c & 31) as u32; // mask to 5 bits
+        let inv = ((32 - k) & 31) as u32;
+        cols.c_masked = Word::from(k);
+        cols.c_inverse = Word::from(inv);
+
+        // --- 2) Instruction selectors.
+        cols.is_rotl = F::from_bool(event.code == Opcode::I32Rotl.code());
+        cols.is_rotr = F::from_bool(event.code == Opcode::I32Rotr.code());
+
+        // --- 3) Decomposition: precompute the two contributions (left/right) from the executor.
+        // For real rows, we will constrain in AIR that: a = left OR right and (left & right) == 0.
+        // For ROTR(k): left = b << (32 - k), right = b >> k
+        // For ROTL(k): left = b >> (32 - k), right = b << k
+        let (left_bytes, right_bytes) = if event.code == Opcode::I32Rotl.code() {
+            (
+                event.b.wrapping_shr(inv).to_le_bytes(), // b >> (32 - k)
+                event.b.wrapping_shl(k).to_le_bytes(),   // b << k
+            )
+        } else {
+            (
+                event.b.wrapping_shl(inv).to_le_bytes(), // b << (32 - k)
+                event.b.wrapping_shr(k).to_le_bytes(),   // b >> k
+            )
+        };
+
+        cols.left_shifted = Word(left_bytes.map(F::from_canonical_u8));
+        cols.right_shifted = Word(right_bytes.map(F::from_canonical_u8));
+
+        // --- 4) Byte lookups: a = left OR right (byte-wise).
+        for ((a_b, l_b), r_b) in event.a.to_le_bytes().into_iter().zip(left_bytes).zip(right_bytes)
```

Comment:
you can use [izip!](https://docs.rs/itertools/latest/itertools/macro.izip.html) here